### PR TITLE
Implemented ReflectionProperty::setAccessible(), getValue() and setValue() and ReflectionClass::getStaticProperties()

### PR DESCRIPTION
**Pull request by kukulich**

### CODE REVIEW

File: src/Reflection/ReflectionClass.php [hunk 1]
```diff
@@ -1287,36 +1289,28 @@ public function __clone()
...
         if ( ! $property || ! $property->isStatic()) {
-            throw new PropertyDoesNotExist('Property does not exist on class or is not static');
-        }
-
-        // PHP behaviour is to simply say "property does not exist" if accessing
-        // protected or private values. Here we be a little more explicit in
-        // reasoning...
-        if ( ! $property->isPublic()) {
-            throw new PropertyNotPublic('Property is not public');
+            throw PropertyDoesNotExistOrIsNotStatic::fromName($propertyName);
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: You know if it does not exist or if it is static, so throwing a `PropertyDoesNotExistOrIsNotStatic` feels wrong.

A better way to handle this is having `PropertyDoesNotExistOrIsNotStatic` be an abstract class, and then have two subtypes for the actual failure: that would make it much easier to deal with for a consumer, and would read more sensibly.

File: src/Reflection/Adapter/ReflectionClass.php [hunk 2]
```diff
@@ -400,23 +403,36 @@ public function isSubclassOf($class)
...
     public function getStaticPropertyValue($name, $default = null)
     {
-        throw new Exception\NotImplemented('Not implemented');
+        try {
+            return $this->betterReflectionClass->getStaticPropertyValue($name);
+        } catch (PropertyDoesNotExistOrIsNotStatic $e) {
+            if (2 === \func_num_args()) {
+                return $default;
+            }
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionClass.php]: Spacing please: spaces around logical blocks :-)

File: src/Reflection/Adapter/ReflectionClass.php [hunk 3]
```diff
@@ -400,23 +403,36 @@ public function isSubclassOf($class)
...
+        } catch (Throwable $e) {
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionClass.php]: When can this happen?

File: src/Reflection/Adapter/ReflectionClass.php [hunk 4]
```diff
@@ -400,23 +403,36 @@ public function isSubclassOf($class)
...
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionClass.php]: Shouldn't we populate the previous exception?

File: src/Reflection/Adapter/ReflectionClass.php (hunk 2 above)
Comment (Ocramius) [reply src/Reflection/Adapter/ReflectionClass.php]: Shouldn't we populate the previous exception?

File: src/Reflection/Adapter/ReflectionClass.php [hunk 5]
```diff
@@ -400,23 +403,36 @@ public function isSubclassOf($class)
...
     /**
      * {@inheritDoc}
      */
     public function setStaticPropertyValue($name, $value)
     {
-        throw new Exception\NotImplemented('Not implemented');
+        try {
+            $this->betterReflectionClass->setStaticPropertyValue($name, $value);
+        } catch (Throwable $e) {
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionClass.php]: Shouldn't we populate the previous exception?

File: src/Reflection/Adapter/ReflectionObject.php [hunk 6]
```diff
@@ -370,23 +373,36 @@ public function isSubclassOf($class)
...
     public function getStaticPropertyValue($name, $default = null)
     {
-        throw new Exception\NotImplemented('Not implemented');
+        try {
+            return $this->betterReflectionObject->getStaticPropertyValue($name);
+        } catch (PropertyDoesNotExistOrIsNotStatic $e) {
+            if (2 === \func_num_args()) {
+                return $default;
+            }
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionObject.php]: Spacing

File: src/Reflection/Adapter/ReflectionObject.php [hunk 7]
```diff
@@ -370,23 +373,36 @@ public function isSubclassOf($class)
...
+        } catch (Throwable $e) {
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionObject.php]: Shouldn't we populate the previous exception?

File: src/Reflection/Adapter/ReflectionObject.php (hunk 6 above)
Comment (Ocramius) [reply src/Reflection/Adapter/ReflectionObject.php]: Shouldn't we populate the previous exception?

File: src/Reflection/Adapter/ReflectionObject.php [hunk 8]
```diff
@@ -370,23 +373,36 @@ public function isSubclassOf($class)
...
     /**
      * {@inheritDoc}
      */
     public function setStaticPropertyValue($name, $value)
     {
-        throw new Exception\NotImplemented('Not implemented');
+        try {
+            $this->betterReflectionObject->setStaticPropertyValue($name, $value);
+        } catch (Throwable $e) {
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionObject.php]: Shouldn't we populate the previous exception?

File: src/Reflection/Adapter/ReflectionProperty.php [hunk 9]
```diff
@@ -47,15 +51,27 @@ public function getName()
      */
     public function getValue($object = null)
     {
-        throw new Exception\NotImplemented('Not implemented');
+        try {
+            return $this->betterReflectionProperty->getValue($object);
+        } catch (NoObjectProvided | NotAnObject $e) {
+            return null;
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionProperty.php]: I think this was fixed somewhen... 7.1 maybe?

File: src/Reflection/Adapter/ReflectionProperty.php [hunk 10]
```diff
@@ -47,15 +51,27 @@ public function getName()
...
+        } catch (Throwable $e) {
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionProperty.php]: Shouldn't we populate the previous exception?

File: src/Reflection/Adapter/ReflectionProperty.php [hunk 11]
```diff
@@ -47,15 +51,27 @@ public function getName()
...
      */
     public function setValue($object, $value = null)
     {
-        throw new Exception\NotImplemented('Not implemented');
+        try {
+            $this->betterReflectionProperty->setValue($object, $value);
+        } catch (NoObjectProvided | NotAnObject $e) {
+            return null;
+        } catch (Throwable $e) {
+            throw new CoreReflectionException($e->getMessage());
```
Comment (Ocramius) [inline src/Reflection/Adapter/ReflectionProperty.php]: Shouldn't we populate the previous exception?

File: src/Reflection/ReflectionProperty.php [hunk 12]
```diff
@@ -361,4 +372,120 @@ public function __clone()
...
+     * @throws ClassDoesNotExist
+     * @throws NoObjectProvided
+     * @throws NotAnObject
+     * @throws ObjectNotInstanceOfClass
+     */
+    public function getValue($object = null)
+    {
+        $this->assertAccessibility();
+
+        $declaringClassName = $this->getDeclaringClass()->getName();
```
Comment (Ocramius) [inline src/Reflection/ReflectionProperty.php]: Scenario to check: https://3v4l.org/8fBDe

File: src/Reflection/ReflectionProperty.php [hunk 13]
```diff
@@ -361,4 +372,120 @@ public function __clone()
...
+     * @return mixed
+     * @throws PropertyNotAccessible
+     * @throws ClassDoesNotExist
+     * @throws NoObjectProvided
+     * @throws NotAnObject
+     * @throws ObjectNotInstanceOfClass
+     */
+    public function getValue($object = null)
+    {
+        $this->assertAccessibility();
```
Comment (Ocramius) [inline src/Reflection/ReflectionProperty.php]: IMO, our own implementation should drop the `setAccessible()` bullcrap

File: src/Reflection/ReflectionProperty.php [hunk 14]
```diff
@@ -361,4 +372,120 @@ public function __clone()
...
+
+        $declaringClassName = $this->getDeclaringClass()->getName();
+
+        if ($this->isStatic()) {
+            $this->assertClassExist($declaringClassName);
+
+            return Closure::bind(function (string $propertyName) {
+                return self::${$propertyName};
```
Comment (Ocramius) [inline src/Reflection/ReflectionProperty.php]: Instead of `self::`, use `$declaringClassName`

File: src/Reflection/ReflectionProperty.php [hunk 15]
```diff
@@ -361,4 +372,120 @@ public function __clone()
...
+            }, null, $declaringClassName)->__invoke($this->getName());
+        }
+
+        $this->assertObject($object);
+
+        return Closure::bind(function (string $propertyName) {
+            return $this->{$propertyName};
```
Comment (Ocramius) [inline src/Reflection/ReflectionProperty.php]: Instead of `$this`, use a parameter `$object` and call the closure with that

Comment (Ocramius) [reply src/Reflection/ReflectionProperty.php]: (this is to confuse the static analysers *less*

File: src/Reflection/Adapter/ReflectionProperty.php (hunk 9 above)
Comment (kukulich) [reply src/Reflection/Adapter/ReflectionProperty.php]: Nope, it throws a warning and returns `NULL`.

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (kukulich) [reply src/Reflection/ReflectionClass.php]: Ok. I just used the code that was already here :) https://github.com/Roave/BetterReflection/blob/master/src/Reflection/ReflectionClass.php#L1309

File: src/Reflection/Adapter/ReflectionProperty.php (hunk 9 above)
Comment (Ocramius) [reply src/Reflection/Adapter/ReflectionProperty.php]: Dang :-\

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (Ocramius) [reply src/Reflection/ReflectionClass.php]: Aye - I know my comments sometimes sound annoying and out of context, but I really just see the diffs only ^_^

Comment (kukulich) [reply src/Reflection/ReflectionClass.php]: It's ok :) I hope I will someday prepare a pull request that can be merged straight away :)

File: src/Reflection/ReflectionProperty.php (hunk 13 above)
Comment (kukulich) [reply src/Reflection/ReflectionProperty.php]: Done. `setAccessible()` implemented only in adapters.

**Comment by kukulich** [comment]
@Ocramius I hope everything is ok now :)

**Comment by Ocramius** [comment]
@kukulich merged, thanks! I tried to improve something in https://github.com/Roave/BetterReflection/commit/9e82778965aa1177d2bd7936b9314e8be5a62ebf, but failed due to how the test is (reasonably) designed.

:ship: and thanks!

**Comment by kukulich** [comment]
@Ocramius I'm only on mobile now but if I understand it well then remove this line https://github.com/Roave/BetterReflection/blob/master/test/unit/Reflection/Adapter/ReflectionClassTest.php#L90 and implement special test for "getStaticProperties". Some other methods are already tested in similar way.

**Comment by Ocramius** [comment]
That's alright: the naming is not the end of the world here, as it would
probably be more confusing to have a different one

Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

**Pull request merged**